don't collect observed `ArraySymbolics`

hexaeder opened the pull request:
If `is_observed` returns true for `ArraySymbolic`, forward directly to `observed` instead of collecting first.

## Checklist

- [ ] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API

AayushSabharwal reviewed (commented):
Looks good! Just needs a minor change and some tests. Thanks for the PR

hexaeder commented:
I removed the tuple stuff and added some tests to ensure that `ArraySymbolics` are forwarede "as is" without collecting if `is_(parameter/variable/observed)` returns true.

AayushSabharwal commented:
I just ran into some edge case that is tested in SciMLBase downstream in my own work. It will be broken by the changes in this PR. Could you refactor the check in `_getp` so that instead of checking `is_observed(sys, p)` it checks `any(x -> is_observed(sys, x), collect(p))`? Also, it looks like the downstream environment needs to cap SymbolicUtils to `<1.6`. Please add that here as well so that CI runs.

hexaeder commented:
> Could you refactor the check in `_getp` so that instead of checking `is_observed(sys, p)` it checks `any(x -> is_observed(sys, x), collect(p))`?

I guess you're referring to `_getu` as I am not altering `_getp` at all? I think changing to `any(x -> is_observed(sys, x), collect(p))` is kinda the opposite of what this PR is trying to achieve. Without this PR:

- user requests `ArraySymbolic` `sym`,
  - if `is_variable(sym)` -> call `variable_index(sym)` without collecting
  - if `is_parameter(sym)` -> call `parameter_index(sym)` without collecting
  - else collect and try again on the `::NotSymbolic` Array of `ScalarSymbolic`

My PR tries to apply the same behavior to `observed` states, meaning that `is_observed(sym)` returns true, call `observed(sym)` directly **without** collecting `sym`. If `is_observed(sym)` is `false`, it will collect and call it again on the collected array, which will then hit the code path which does more or less exactly what you suggested anyway

https://github.com/SciML/SymbolicIndexingInterface.jl/blob/c9c7b6c1fd6fd3ac787e6a9d554d84efcd80f2e4/src/state_indexing.jl#L147-L170

So I am not sure what happens in your edge cases, but maybe `is_observed` does not fall back to some `is_observed(_::ArraySymbolic) = false` even though `observed` cannot handle the `ArraySymbolic`?

AayushSabharwal commented:
The problem is that if you have an `ODESystem` with parameters `p[1], p[2], p[3]` (all scalarized, so they're stored as 3 scalars instead of a single vector) and try to do `getu(sys, p[2:3])`, `is_parameter(sys, p[2:3])` is `false`, but `is_observed(sys, p[2:3])` is `true`, which means that it will generate a function for `p[2:3]`. These generated functions break type inference, so the operation becomes type-unstable. It's also unnecessarily expensive when in reality this should just be a couple of `getindex` calls.

I'll check this case with your PR shortly, since it might just not break inference in which case it's mostly fine.
